platform/linux-generic/include/odp_ipsec_internal.h

```diff
@@ -158,6 +158,8 @@ struct ipsec_sa_s {
 	void		*context;
 	odp_queue_t	queue;
 
+	/* ESN higher 32 bits length */
+	uint32_t	seqh_len;
```

Comment:
I think the comment is slightly misleading as the "length" of the 32 high bits of the sequence number is of course 32 bits, i.e. 4 bytes.

I would actually suggest removing this field and adding a "combined_mode" or "aead" flag instead. The code could then check the flag when it needs to know whether to put the high part of the seqno after the next header field or not. And for the number of bytes the code could just use the constant 4 (defined appropriately).

Reply:
>I would actually suggest removing this field and adding a "combined_mode" or "aead" flag instead.
Agree with your suggestion.

  >And for the number of bytes the code could just use the constant 4 (defined appropriately).
Sure, will use an appropriate macro.

Reply:
As per your suggestion added a new "combined_mode". But realized that removal of "seqh_len" field will introduce extra "if" checks in the data path, so had to keep it.

Reply:
> removal of "seqh_len" field will introduce extra "if" checks in the data path, so had to keep it.

The thing is that passing segh_len instead of the constant 4 to things like odp_packet_copy_from_mem() prevents optimizations by the compiler. Since the compiler cannot see that the value is always 4 (if you additionally skip the calls if it is zero), it must create a loop for the memcpy etc. So my suggestion really was about having  an SA flag that tells whether the high order bits of the sequence number are to be put in the packet after the ESP trailer for the purpose of ICV computation. Then the code that already checks the flag could use a compile time constant for the "length".

In a couple of places having segh_len still around may make sense to get rid of an unpredictable branch. But maybe even in those places you could just compute the "length" based on the flag value. (i.e. segh_len = flag * 4 or segh_len = (1- flag) * 4, depending on the definition of the flag). Such computation should be put in a function.

Reply:
Thanks for the giving deeper insight, I will try to modify this as below:

1. For the constant 4, will use this macro in places like "odp_packet_copy_from_mem()".
/* ESN higher bits length in bytes */
#define IPSEC_SEQH_LEN      4

2. For "length" at other places will use below function to compute its value:
static inline uint32_t ipsec_get_seqh_len(ipsec_sa_t *ipsec_sa)
{
        if (!ipsec_sa->icv_len || ipsec_sa->combined_mode) {
                return 0;
        else
               return IPSEC_SEQH_LEN;
}

3. Will remove the "seqh_len".

Reply:
That looks a bit awkward and probably results in the branches that you wanted to avoid. How about defining insert_seq_hi flag (or whatever it would be named) in IPsec SA so that the flag would tell exactly when high part of the seqno is inserted after the next header field. Then you would not need to special case the null algorithm using the icv lenght. And this function would not need any conditionals but could simply return ipsec_sa->insert_seq_hi * IPSEC_SEQ_HI_LEN.

Reply:
Agreed.
1. /* ESN higher bits length in bytes */
#define IPSEC_SEQH_LEN 4
2. Will rename the "seqh_len" to  this new flag "insert_seq_hi".
3. Will set this new flag during SA creation.
```
        ipsec_sa->insert_seq_hi = (ipsec_sa->esn) ? 1 : 0;
        if (!ipsec_sa->icv_len || ipsec_sa->combined_mode)
                ipsec_sa->insert_seq_hi = 0;

```

4. New function does not have conditions.

```
static inline uint32_t ipsec_get_seqh_len(ipsec_sa_t *ipsec_sa)
{
        return (ipsec_sa->insert_seq_hi * IPSEC_SEQ_HI_LEN);
}

```